oauthex/oauthex.go

```diff
@@ -5,88 +5,38 @@
-	// This value may be internationalized.
-	ResourceName string `json:"resource_name,omitempty"`
-
-	// ResourceDocumentation (resource_documentation) is an optional URL of a page containing
-	// human-readable information for developers using the protected resource.
-	// This value may be internationalized.
-	ResourceDocumentation string `json:"resource_documentation,omitempty"`
-
-	// ResourcePolicyURI (resource_policy_uri) is an optional URL of a page containing
-	// human-readable policy information on how a client can use the data provided.
-	// This value may be internationalized.
-	ResourcePolicyURI string `json:"resource_policy_uri,omitempty"`
-
-	// ResourceTOSURI (resource_tos_uri) is an optional URL of a page containing the protected
-	// resource's human-readable terms of service. This value may be internationalized.
-	ResourceTOSURI string `json:"resource_tos_uri,omitempty"`
-
-	// TLSClientCertificateBoundAccessTokens (tls_client_certificate_bound_access_tokens) is an
-	// optional boolean indicating support for mutual-TLS client certificate-bound
-	// access tokens (RFC 8705). Defaults to false if omitted.
-	TLSClientCertificateBoundAccessTokens bool `json:"tls_client_certificate_bound_access_tokens,omitempty"`
-
-	// AuthorizationDetailsTypesSupported (authorization_details_types_supported) is an optional
-	// slice of 'type' values supported by the resource server for the
-	// 'authorization_details' parameter (RFC 9396).
-	AuthorizationDetailsTypesSupported []string `json:"authorization_details_types_supported,omitempty"`
+// GetAuthServerMeta retrieves OAuth 2.0 Authorization Server Metadata for the given issuer URL.
+// This forwards to the internal implementation.
+func GetAuthServerMeta(ctx context.Context, issuerURL string, c *http.Client) (*AuthServerMeta, error) {
+	return oauthex.GetAuthServerMeta(ctx, issuerURL, c)
+}
 
-	// DPOPSigningAlgValuesSupported (dpop_signing_alg_values_supported) is an optional
-	// slice of JWS signing algorithms supported by the resource server for validating
-	// DPoP proof JWTs (RFC 9449).
-	DPOPSigningAlgValuesSupported []string `json:"dpop_signing_alg_values_supported,omitempty"`
+// RequirePKCE checks that the authorization server for issuerURL supports PKCE,
+// by verifying that CodeChallengeMethodsSupported is non-empty.
+// It returns an error if metadata cannot be fetched or PKCE is not advertised.
+func RequirePKCE(ctx context.Context, issuerURL string, c *http.Client) error {
+	asm, err := oauthex.GetAuthServerMeta(ctx, issuerURL, c)
+	if err != nil {
+		return err
+	}
 
-	// DPOPBoundAccessTokensRequired (dpop_bound_access_tokens_required) is an optional boolean
-	// specifying whether the protected resource always requires the use of DPoP-bound
-	// access tokens (RFC 9449). Defaults to false if omitted.
-	DPOPBoundAccessTokensRequired bool `json:"dpop_bound_access_tokens_required,omitempty"`
+	if len(asm.CodeChallengeMethodsSupported) == 0 {
```

Comment:
Don't we also have to check code_verifier? https://datatracker.ietf.org/doc/html/draft-ietf-oauth-v2-1-13#section-7.5.2

Reply:
The draft spec says that only code challenge methods needs to be checked:
https://github.com/modelcontextprotocol/modelcontextprotocol/blob/main/docs/specification/draft/basic/authorization.md, the section entitled Authorization Code Protection.